### PR TITLE
WT-12861 consolidate cache config validation into a single function

### PR DESCRIPTION
**Pull request by bzm-jas**

### CODE REVIEW

**Review by etienneptl — COMMENTED** [review]
Nice one! Left a few comments.

File: src/conn/conn_cache.c
Comment (sueloverso) [inline src/conn/conn_cache.c:65]: Isn't this already done from `wti_debug_mode_configuration` and, instead of needing the boolean, you can just use `FLD_ISSET(conn, WT_CONN_DEBUG_CONFIGURATION)`? It is better to have one place getting the configuration.

File: src/conn/conn_cache.c
Comment (sueloverso) [inline src/conn/conn_cache.c:81]: It is confusing to see that `cache->eviction_target` was just checked above and now we're setting it. Maybe add a comment after the conditional that now we're resetting the values via reconfigure or restart or however we get here.

File: src/conn/conn_cache.c
Comment (sueloverso) [inline src/conn/conn_cache.c:84]: If someone can reconfigure all the values, don't we need to repeat all the same invalid checks we did on `create`? For example, if one can reconfigure both `cache->eviction_target` and `cache->eviction_trigger` don't we need to check like at line 69? The checks below do not repeat the checks above but do potentially change values in an invalid way.

If so, then this code should be re-arranged to have one set of those EINVAL conditionals.

File: src/conn/conn_cache.c
Comment (etienneptl) [reply src/conn/conn_cache.c:65]: The issue is that the debug flags are set after setting the cache. @bzm-jas posted a comment about it in the JIRA ticket [here](https://jira.mongodb.org/browse/WT-12861?focusedCommentId=6395743&focusedId=6395743&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6395743).
I asked in our team channel if there was any specific reason why we do it in this order. The current code seems to indicate this order is required but I don't understand why and I wonder if this comment was valid before but now now:
```
    /*
     * This function expects the cache to be created so parse this after the rest of the connection
     * is set up.
     */
    WT_ERR(__wti_debug_mode_config(session, cfg));
```

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: Good idea, added in: [d2ff734](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/d2ff7347ce5170f632ac585e2200d712789759cf)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: it was for the unit test, now removed: [820f876](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/820f876e70fd17482532ce144839326bd7101586)

File: src/include/extern.h
Comment (bzm-jas) [reply src/include/extern.h]: removed here: [820f876](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/820f876e70fd17482532ce144839326bd7101586)

File: test/suite/test_config12.py
Comment (bzm-jas) [reply test/suite/test_config12.py:77]: sounds good, comment added in: [ac8d882](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/ac8d8825adee297c35f5bceaa8d840a58353fcc0)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c:54]: Thanks for the suggestion, Sue! I will be renaming the function for better clarity.

Regarding splitting the functions: the original purpose of this ticket was to change the logic for checking values in WT_CACHE so they would only be set after validation. In other words, for invalid configurations, they would not be set, and the function would fail immediately.

However, since we need to keep auto-fixing for now, extracting the values and setting them in WT_CACHE is the best approach. In the future, we do want to progressively move to the original idea. In that case, if we have a function that extracts and sets the cache values, it would need to be modified to include a temporary structure for storing the config values to validate and another function for setting all valid configs to the WT_CACHE structure of the current session. So it will be changed entirely.

Since there could be major changes to this part of the code, I'm open to discussing the best way to structure it now to benefit and simplify future modifications. I could also rename the function to reflect both extraction and validation, like `__conn_cache_config_prepare` or `__conn_cache_config_setup`. Please let me know what you think!

File: src/conn/conn_cache.c
Comment (etienneptl) [inline src/conn/conn_cache.c:70]: For each message, should we reuse the **exact same name** as the configuration? This way users will know what to change.
For instance:
```suggestion
            WT_RET_MSG(session, EINVAL, "eviction_target must be lower than the eviction_trigger");
```

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c:82]: sounds good, comment added in: [c7c0742](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/c7c074299c46b9cb2a5ceb857089d5053db06b33)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: good point, I've modified the warning messages in: [feb593b](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/feb593b0f3e88fc188dd0cc5c8b7725602d551f0)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c:85]: I agree that both checks should be performed, but using EINVAL for all checks causes a lot of failures, I will look into finding an approach to log these messages without breaking our tests.

File: src/conn/conn_cache.c
Comment (ajmorton) [inline src/conn/conn_cache.c:106]: General wording suggestion for this and the other warning messages
```suggestion
              "config eviction_dirty_target (%f) cannot exceed than eviction_target (%f). Setting"
```

File: src/conn/conn_cache.c
Comment (ajmorton) [inline src/conn/conn_cache.c:136]: `DBL_EPSILON` is an internal variable name. We should avoid displaying it to the end user:
```suggestion
              "config eviction_updates_target (%f) cannot be zero. Setting to 50% of eviction_updates_target (%f)",
```

File: test/suite/test_config12.py
Comment (ajmorton) [inline test/suite/test_config12.py:40]: Perhaps we can define this function inside `WiredTigerTestCase` instead of having two definitions here and in `test_verbose01.py`. 
@etienneptl you wrote the original `test_verbose01.py`. Was there a reason we defined expect_verbose in `test_verbose01.py`?

File: test/suite/test_config12.py
Comment (ajmorton) [inline test/suite/test_config12.py:80]: Can you add some more tests cases for non-default configs? For example we're currently not testing cases like `cache->eviction_dirty_trigger > cache->eviction_trigger`

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: fixed here: [9e4baa7](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/9e4baa7696ed61b4507e4766121155606277fcd8)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: Thanks for picking this up, I've modified `__validate_cache_config` so now the values gets changed to percentage straight away after retrieving and setting the value. See this commit: [9e4baa7](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/9e4baa7696ed61b4507e4766121155606277fcd8)
I'd prefer to change the way we access `conn->cache` in another ticket because I think there's already quite a bit of changes going on for this PR, it might be clearer to do for a seperate ticket.

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: Comment rephrased in: [9e4baa7](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/9e4baa7696ed61b4507e4766121155606277fcd8)

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: Comment rephrased in: [9e4baa7](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/9e4baa7696ed61b4507e4766121155606277fcd8)

File: test/suite/test_config12.py
Comment (bzm-jas) [reply test/suite/test_config12.py:80]: Sure, I will add at least one test for each of the checks.

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c:78]: I've removed the `create=true` logic and the second call to` __validate_cache_config(create=true)` in: [9e4baa7](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/9e4baa7696ed61b4507e4766121155606277fcd8)

File: src/conn/conn_cache.c
Comment (etienneptl) [reply src/conn/conn_cache.c:65]: I wish we didn't have to read the configuration twice and call `__wti_debug_mode_config` beforehand but if that's the only solution, I'd be happy with that as well!

File: test/suite/test_config12.py
Comment (etienneptl) [reply test/suite/test_config12.py:40]: Simply because it was the only test using that function and it was designed to meet the needs of `verbose01`. This version is slightly different than the one defined in `verbose01` but I agree that it should be accessible from any test, it can be useful.

File: test/suite/test_config12.py
Comment (etienneptl) [inline test/suite/test_config12.py:77]: Can we define the second arg as a variable?

Comment (bzm-jas) [reply test/suite/test_config12.py]: yep, defined as a variable here: [5f55183](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/5f551837bf0db3f86bfc0e25c008057f1655f23c)

File: test/suite/test_config12.py
Comment (bzm-jas) [reply test/suite/test_config12.py:83]: I have added more tests in: [390b88b](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/390b88bcfae2f88596548c7296ee1ad12c5c994a)

File: dist/api_data.py
Comment (bzm-jas) [reply dist/api_data.py:580]: Currently there's no plans in extending this functionality, if we are not considering to extend it, I'm happy to rename the flag.

File: src/conn/conn_cache.c
Comment (marcbutler) [inline src/conn/conn_cache.c:123]: I'm wondering if the creation of macro for outputting the debug might make these corrective statements clearer, so a macro like:
```
#define CONFIG_DEBUG(msg, ...) \
  check if debugging is enabled and if so output msg
```
This would transform the code to something like:

```suggestion
    if (cache->eviction_dirty_target > cache->eviction_target) {
        cache->eviction_dirty_target = cache->eviction_target;
        CONFIG_DEBUG("config eviction_dirty_target=%f cannot exceed eviction_target=%f. Setting "
              "eviction_dirty_target to %f.", cache->eviction_dirty_target, cache->eviction_target, cache->eviction_target);        
    }
```

Note I also moved the change to be close to the condition checked, as the log messages are detailed which is very helpful, they are visually "noisy" in the code. Keeping the evaluation and the modification close to gether helps to clarify the purpose when scanning the code.

**Review by marcbutler — COMMENTED** [review]
Reach out if any clarification is needed.

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c:123]: I think creating a macro for debug is a good idea. If it’s clear enough from a readability standpoint that the messages are only printed when debug mode is enabled then I’m happy to modify it. In terms of moving the modification, we can't move it before printing the message. Otherwise, the invalid config would be overwritten, and the message would be using the post-modified correct config. We can bypass this by storing the previous value but I think it might be a bit unnecessary in this case.

File: src/conn/conn_cache.c
Comment (etienneptl) [inline src/conn/conn_cache.c:10]:
```suggestion
#include "wt_internal.h"

#define WT_CONFIG_DEBUG(session, fmt, ...) \
```

**Review by etienneptl — APPROVED** [review]
LGTM! Thank you for addressing all the feedback.

File: src/conn/conn_cache.c
Comment (ajmorton) [reply src/conn/conn_cache.c]: SGTM! I've created WT-13161 to track this

**Review by ajmorton — APPROVED** [review]
LGTM pending a few small comments. Thanks for all your work!
To be safe, could you also run a full patch before merging this?

File: src/conn/conn_cache.c
Comment (bzm-jas) [reply src/conn/conn_cache.c]: sounds good, comment removed in: [3da403d](https://github.com/wiredtiger/wiredtiger/pull/10548/commits/3da403dac63bb5f6910e92bab4481801e14a83ca)

File: src/conn/conn_cache.c
Comment (marcbutler) [inline src/conn/conn_cache.c:117]: Thanks I find this very readable. Comment only.

**Review by marcbutler — APPROVED** [review]
There are many reviewers, so I just looked at the work I commented on -- thanks.
lgtm

**Review by etienneptl — APPROVED** [review]
LGTM! Great work!

**Pull request merged**
